### PR TITLE
Fix Factors().as_expr(): don't multiply rational exponents

### LINKED ISSUE #18059: Simplify of simple trig expression fails
trigsimp in various versions, including 1.5, incorrectly simplifies cos(x)+sqrt(sin(x)**2) as though it were cos(x)+sin(x) for general complex x. (Oddly it gets this right if x is real.)

Embarrassingly I found this by accident while writing sympy-based teaching material...

### PR DESCRIPTION
**Pull request by gschintgen**
#### References to other Issues or PRs

Fixes #18059 

#### Brief description of what is fixed or changed
This commit fixes an invalid simplification in exprtools.py:

Before:
`Factors(sqrt(x**2)).as_expr() == x`
After:
`Factors(sqrt(x**2)).as_expr() == sqrt(x**2)`

This invalid simplification also led to a wrong result
by fu() (and hence trigsimp and simplify). Issue #18059.

Test cases are added.

#### Release Notes

* core
  * Fixed a bug in `Factors().as_expr()` that led to simplifications that are not generally valid for complex numbers. (Exponents were always multiplied in power-of-a-power situations.)

### CODE REVIEW

File: sympy/core/exprtools.py [hunk 1]
```diff
@@ -448,14 +448,12 @@ def as_expr(self):  # Factors
         args = []
         for factor, exp in self.factors.items():
             if exp != 1:
-                b, e = factor.as_base_exp()
                 if isinstance(exp, int):
```
Comment (jksuom) [inline sympy/core/exprtools.py]: Could `exp` be a SymPy Integer?

Comment (gschintgen) [reply sympy/core/exprtools.py]: Yes and in fact that seems to be the norm (at least for the few simple manual tests that I ran). In that case the general case applies (`args.append(factor**exp)`) and still returns the right result. I must admit though that I'm wondering a bit about that special-casing of the `int` case. Is that only to have it sympified? I'm also unsure about the `_keep_coeff`.
Any further insight is much appreciated.

Comment (smichr) [reply sympy/core/exprtools.py]: `_keep_coeff` is just there to keep 2-arg muls from evaluating (see docstring)

Comment (smichr) [reply sympy/core/exprtools.py]: From a quick look at Factors it seems that all values should be Basic type so the check for `int` should, perhaps, just be `Integer`.

Comment (gschintgen) [reply sympy/core/exprtools.py]: What's strange though is that IIRC codecov marked the code for the `isinstance(exp, int)` case as being covered.
So both cases (int and Integer) should ideally end up using `_keep_coeff`?
I'll change it to `isinstance(exp, (int, Integer))` then.

Comment (smichr) [reply sympy/core/exprtools.py]: But the only place I can see an int potentially creeping in is at lines 360-361 where a default of 0 might be added to something that came from powers dict. If the raw 0 were changed to S.Zero then that would fix that (assuming that `as_powers_dict` might be returning an int. If the int case were fixed then problems like here will be avoided elsewhere.

**Pull request merged**